### PR TITLE
K8s: Add storage dual writer

### PR DESCRIPTION
**Pull request by toddtreece**
**What is this feature?**

Adds a `DualWriter` for writing to both legacy storage and storage that implements `storage.Interface` (etcd for now).

The [README](https://github.com/grafana/grafana/blob/b55fe04a60bc670fbd23b16e7cb8fd87a0b8bc9b/pkg/services/grafana-apiserver/README.md) has more information about how to test this feature.

**Why do we need this feature?**

This is needed as a migration step toward unified storage.

**Who is this feature for?**

@grafana/grafana-app-platform-squad 

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+// - rest.Storage
+// - rest.TableConvertor
+// - rest.Scoper
+// - rest.SingularNameProvider
+//
+// These interfaces are optional, but they all should be implemented to fully support dual writes:
+// - rest.Creater
+// - rest.Updater
+// - rest.GracefulDeleter
+// - rest.CollectionDeleter
```
Comment (toddtreece) [inline pkg/services/grafana-apiserver/rest/dualwriter.go:53]: not sure if this is a good idea, but by making these interfaces optional we can limit the amount of verbs we expose via discovery when dual write is disabled.

for example, with dual write disabled:
```
kubectl get --raw="/apis/playlist.x.grafana.com/v0-alpha" | jq
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "playlist.x.grafana.com/v0-alpha",
  "resources": [
    {
      "name": "playlists",
      "singularName": "playlist",
      "namespaced": true,
      "kind": "Playlist",
      "verbs": [
        "get",
        "list"
      ]
    }
  ]
}
```

with dual write enabled:
```
kubectl get --raw="/apis/playlist.x.grafana.com/v0-alpha" | jq
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "playlist.x.grafana.com/v0-alpha",
  "resources": [
    {
      "name": "playlists",
      "singularName": "playlist",
      "namespaced": true,
      "kind": "Playlist",
      "verbs": [
        "create",
        "delete",
        "deletecollection",
        "get",
        "list",
        "patch",
        "update",
        "watch"
      ]
    }
  ]
}
```

Comment (ryantxu) [reply pkg/services/grafana-apiserver/rest/dualwriter.go:53]: for this to work we will really need everything except "watch" and "deletecollection" in the legacy storage, but I think this is a good idea

**Review by ryantxu — APPROVED** [review]
LGTM -- good first step, but obviouly more needed to actually write!

**Review by charandas — APPROVED** [review]
LGTM!

**Pull request merged**